Document cluster wide rate limits for 23.1 (#1140)
redpanda-data/documentation: docs/manage/cluster-maintenance/manage-throughput.mdx

The properties for node-wide throughput quota balancing are configured at the cluster level, for all nodes in a cluster:

- [kafka_throughput_limit_node_in_bps](../../../reference/cluster-properties#kafka_throughput_limit_node_in_bps): a node's total throughput limit for ingress traffic.
- [kafka_throughput_limit_node_out_bps](../../../reference/cluster-properties#kafka_throughput_limit_node_out_bps): a node's total throughput limit for egress traffic.
- [kafka_quota_balancer_node_period_ms](../../../reference/cluster-properties#kafka_quota_balancer_node_period_ms): the period at which the quota balancer runs to balance throughput quota between a node's shards.

> - [kafka_quota_balancer_min_shard_throughput_ratio](../../../reference/cluster-properties#kafka_quota_balancer_min_shard_throughput_ratio):	the lowest value of the throughput quota a shard can get in the process of quota balancing, expressed as a ratio of the default shard quota. Zero means there is no minimum, while one means no quota can be taken away by the balancer.

Reviewer: > while one means

nit: it's easy to confuse numeral 'one' to pronoun 'one' in this context

Reviewer 2: @dlex Our style is that for 1 - 9, we spell out the numbers. For numbers greater than 9, we use the numeral. This is pretty standard: https://vectorizedio.atlassian.net/wiki/spaces/DOC/pages/92635182/Redpanda+Documentation+Style+Guide#Numbers

- [kafka_quota_balancer_min_shard_throughput_bps](../../../reference/cluster-properties#kafka_quota_balancer_min_shard_throughput_bps):	the lowest value of the throughput quota a shard can get in the process of quota balancing, in bytes per second. Zero means there is no minimum.

:::note notes

> - By default, all of the listed properties are disabled and no throughput limits or quota balancing are applied. You must manually set them to enable throughput limits and quota balancing.

Reviewer: > By default, all of the listed properties are disabled

This is not quite true. By default, only `kafka_throughput_limit_node_in_bps` and `kafka_throughput_limit_node_out_bps` are disabled (and therefore no throughput limits or quota balancing are applied). The default values of the rest are set in a way so that if `kafka_throughput_limit_node_*_bps` is enabled, the TP limiting will start to work immediately, although some fine tuning of other properties may still be required.

---
title: Manage Throughput
---

<head>
    <meta name="title" content="Manage Throughput | Redpanda Docs"/>
    <meta name="description" content="Manage throughput of cluster nodes."/>
</head>

import ManageThroughputIntro from './shared/_intro-manage-throughput.mdx' 

> <ManageThroughputIntro/>

Reviewer 2: Perhaps:
This topic describes how to manage the throughput of Kafka traffic at the cluster level using configurable properties that limit and protect the use of disk and network resources for individual nodes and for an entire cluster. It also explains how to set node-wide throughput limits and quotas for intra-node (inter-shard) traffic.

## Node-wide rate limits

> The network bandwidth and disk utilization of cluster nodes may be overloaded by clients that consume throughput without limits. To prevent resource overloading caused by unconstrained throughput, Redpanda provides runtime configurable properties that limit and balance throughput of Kafka API traffic. 

Reviewer 2: ```suggestion
The network bandwidth and disk utilization of cluster nodes may be overloaded by clients that consume throughput without limits. To prevent resource overloading caused by unconstrained throughput, Redpanda provides runtime-configurable properties that limit and balance throughput of Kafka API traffic. 
```

> Redpanda's [thread-per-core model](../../../get-started/architecture#thread-per-core-model) necessitates inter-shard (intra-node) communication channels. A proper balancing of throughput for all intra-node channels requires throttling down throughput for shards that are experiencing backpressure, while throttling up throughput for shards that aren't experiencing backpressure. 

Reviewer: > inter-shard (intra-node) communication channels. A proper balancing of throughput for all intra-node channels 

This still keeps confusing me if I put myself in the shoes of a customer reading this. If "intra-node channels" are actually "inter-shard channels", assuming that this is an abstraction of a communication that happens between different shards inside a redpanda node, then this has nothing to do with kafka API thoughput limiting. Then, if the sentence goes on about throttling, is that the throttling of inter-shard communications assumingly? But there is no throttling of that kind in redpanda.

It may be easier to replace this paragraph?

Author: It was my (mis)understanding that intra-core == inter-shard channels. I think it's still useful to bridge the high-level thread-per-core model to the per-core quota limits, so I'll rewrite this accurately, without "inter-shard channels."

> To manage the volume of traffic going through a node, Redpanda implements a throughput quota balancer to enforce node-wide throughput rate limits. The throughput quota accounts for all Kafka API traffic going in and out of a node. The balancer has configurable rate limits of total ingress and egress traffic. To distribute the throughput quota, the balancer periodically monitors the throughput rate of a node's shards, and it distributes quota to the shards that can accept it. The balancer determines a dynamically changing default quota per shard. Each shard has a minimum throughput quota value, which is configurable both as a percentage of the default quota and as an absolute rate limit. 

Reviewer 2: ```suggestion
To manage the volume of traffic going through a node, Redpanda implements a throughput quota balancer to enforce node-wide throughput rate limits. The throughput quota accounts for all Kafka API traffic going in and out of a node. The balancer has configurable rate limits for total ingress and egress traffic. To distribute the throughput quota, the balancer periodically monitors the throughput rate of a node's shards, and it distributes quota to the shards that can accept it. The balancer determines a dynamically changing default quota per shard. Each shard has a minimum throughput quota value, which is configurable both as a percentage of the default quota and as an absolute rate limit. 
```

Reviewer 2: Not sure if this changes the meaning but it makes more sense to me worded this way. Of course, your call.

The properties for node-wide throughput quota balancing are configured at the cluster level, for all nodes in a cluster:

> - [kafka_throughput_limit_node_in_bps](../../../reference/cluster-properties#kafka_throughput_limit_node_in_bps): a node's total throughput limit for ingress traffic.

Reviewer 2: Wondering if these might be easier to view/consume in table format? Thoughts?

Author: Will create tables

- [kafka_throughput_limit_node_out_bps](../../../reference/cluster-properties#kafka_throughput_limit_node_out_bps): a node's total throughput limit for egress traffic.
- [kafka_quota_balancer_node_period_ms](../../../reference/cluster-properties#kafka_quota_balancer_node_period_ms): the period at which the quota balancer runs to balance throughput quota between a node's shards.
- [kafka_quota_balancer_min_shard_throughput_ratio](../../../reference/cluster-properties#kafka_quota_balancer_min_shard_throughput_ratio):	the lowest value of the throughput quota a shard can get in the process of quota balancing, expressed as a ratio of the default shard quota. If set as `0`, there is no minimum, and if set as `1`, no quota can be taken away by the balancer.
- [kafka_quota_balancer_min_shard_throughput_bps](../../../reference/cluster-properties#kafka_quota_balancer_min_shard_throughput_bps):	the lowest value of the throughput quota a shard can get in the process of quota balancing, in bytes per second. Zero means there is no minimum.

> - [max_kafka_throttle_delay_ms](../../../reference/tunable-properties#max_kafka_throttle_delay_ms): when throttling down throughput for a shard, this is the maximum duration that the balancer withholds allocating quota to it.

Reviewer: > this is the maximum duration that the balancer withholds allocating quota to it

I'm sorry but I could not understand what that means.

Throttling (down) is implemented as a delay introduced in the data path of API requests. If the current thoughput exceeds the current quota on a shard, processing of further requests will be delayed so that the throughput is kept within quota limits. This parameter sets the maximum duration for such delays. It exists because cleints may decide that a request has timed out if it has been delayed for too long; so for example if the request timeout period is 30s from client's perspective, it makes sense to set this parameter to e.g. 20s, so that the timeout is never hit.

- [kafka_quota_balancer_window_ms](../../../reference/cluster-properties#kafka_quota_balancer_window_ms): the time window the balancer uses to average the current throughput measurement.

:::note notes

> - By default, both kafka_throughput_limit_node_in_bps and kafka_throughput_limit_node_out_bps are disabled, so no throughput limits or quota balancing are applied. You must manually set them to turn on quota balancing.

Reviewer 2: ```suggestion
- By default, both `kafka_throughput_limit_node_in_bps` and `kafka_throughput_limit_node_out_bps` are disabled, so no throughput limits or quota balancing are applied. You must manually set them to turn on quota balancing.
```

> - By default, both kafka_throughput_limit_node_in_bps and kafka_throughput_limit_node_out_bps are disabled, so no throughput limits or quota balancing are applied. You must manually set them to turn on quota balancing.

Reviewer: > so no throughput limits or quota balancing are applied

It's safer to say only "no throughput limits are applied". Technically, the balancer still runs, although that does not matter if the limits are not applied.

> - kafka_quota_balancer_min_shard_throughput_bps doesn't override the kafka_throughput_limit_node_in_bps and kafka_throughput_limit_node_out_bps limit settings. Consequently, the value of 

Reviewer 2: ```suggestion
- `kafka_quota_balancer_min_shard_throughput_bps` doesn't override the `kafka_throughput_limit_node_in_bps` and `kafka_throughput_limit_node_out_bps` limit settings. Consequently, the value of 
```

---
title: Manage Throughput
---

<head>
  <meta name="title" content="Manage Throughput | Redpanda Docs"/>
  <meta name="description" content="Manage the throughput of Redpanda cluster nodes with configurable properties."/>
</head>

import ManageThroughputIntro from "./shared/_intro-manage-throughput.mdx";

<ManageThroughputIntro/>

## Node-wide throughput limits

> The network bandwidth and disk utilization of cluster nodes may be overloaded by clients that consume throughput without limits. To prevent resource overloading caused by unconstrained throughput, Redpanda provides runtime-configurable properties that limit and balance throughput of Kafka API traffic.
> 
> With Redpanda's [thread-per-core model](../../../get-started/architecture#thread-per-core-model), the Kafka API traffic to and from a client connection is processed by a single core (shard). When a shard handles Kafka traffic of increasing backpressure, it can avoid the negative impact of the backpressure on node resources like memory, storage, and compute.
> 
> To manage the volume of traffic going through a node, Redpanda implements a throughput quota balancer to enforce node-wide throughput rate limits. The throughput quota accounts for all Kafka API traffic going in and out of a node, with the value of quota representing the allowed rate of data passing through in one direction. The balancer has configurable rate limits for total ingress and egress traffic. To distribute the throughput quota, the balancer periodically monitors the throughput rate of a node's shards, and it distributes quota to the shards that can accept it. Each shard has a minimum throughput quota value, which is configurable both as a percentage of the default quota and as an absolute rate limit. The balancer implements its throttler with a dynamically updated delay that is added to the data path of Kafka API requests. 

Reviewer: ```suggestion
The network bandwidth and disk utilization of cluster nodes may be overloaded by clients that produce or consume throughput without limits. To prevent resource overloading caused by unconstrained throughput, Redpanda provides runtime-configurable properties that limit and balance throughput of Kafka API traffic.

To manage the volume of traffic going through a node, Redpanda implements throughput quotas on ingress and egress side of every node. The throughput quota accounts for all Kafka API traffic going in or out of a node, with the value of quota representing the allowed rate of data passing through in one direction. When a connection is in breach of the quota, the throttler advises the client about the delay that would bring the rate back to the allowed level, and implements that delay before handling Kafka API requests. There are configurable rate limits for node total ingress and egress traffic to control the quotas.

With Redpanda's [thread-per-core model](../../../get-started/architecture#thread-per-core-model), the Kafka API traffic to and from a client connection is processed by a single core (shard). In order to manage thoughput quotas efficiently, node quotas are distributed between shards, and each per-shard quota is in turn shared by all connections served by the shard. The job of splitting node quota between shards in an optimal way is done behind the scenes by the component called quota balancer.

To distribute the node throughput quota, the balancer periodically monitors the throughput rate of a node's shards, and it distributes more quota to the shards that can make better use of it than the others. Each shard has a minimum throughput quota value, which is configurable both as a percentage of the default quota and as an absolute rate limit. 
```

---

<head>

>   <meta name="title" content="Manage Throughput | Redpanda Docs" />

Author: Will fix
